### PR TITLE
764 abstract components

### LINKED ISSUE #764: Abstract out runtime component pieces to enable better diagnostics
**What Needs to be Done?**

Make general interfaces for components so diagnostics can be collected.

### PR DESCRIPTION
**Pull request by lifflander**
Fixes #764

- Create a new abstract interface for VT components, universalizing `initialize` and `finalize`.
- Add new type-level dependency management between components to ensure valid startup/finalization order
  - Use topological sort to determine valid order
  - Note: the order is not necessary deterministic across platforms/compilers, but is always the same across all nodes for a given binary and always respects declared dependences
- Create abstraction for `PollableComponent<T>` inheriting from `Progressable` so scheduler knows which components to poll automatically
- Stub abstraction for diagnostics (eventually with have a universal interface for dumping state and calculating diagnostics values at shutdown, e.g., largest or average message size)
  - **Diagnostics will arrive in another PR**
- Stub abstraction for universal buffering for VT components
  - **Universal buffering will arrive in another PR**
- Add base component overload for revealing a name `std::string` for each component for use in debugging and later diagnostics

TODO:
  - [x] Add doxygen
  - [x] Cleanup runtime
  - [x] Hookup `Progressable` to scheduler component

### CODE REVIEW

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/messaging/active.cc  2
- tests/unit/memory/test_memory_active.cc  2
- src/vt/worker/worker_group.h  1
- src/vt/trace/trace.cc  2
- src/vt/worker/worker_group_omp.h  1
- src/vt/sequence/sequencer.impl.h  3
- src/vt/pool/pool.cc  2
- src/vt/messaging/active.h  4
- src/vt/parameterization/parameterization.h  2
- src/vt/runtime/runtime.cc  2
- src/vt/rdma/rdma.h  6
- src/vt/trace/trace.h  2
- tests/unit/memory/test_memory_lifetime.cc  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5390178&bid=17670967)

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- src/vt/messaging/active.h  4
- src/vt/trace/trace.cc  2
- tests/unit/memory/test_memory_active.cc  2
- src/vt/worker/worker_group_omp.h  1
- tests/unit/memory/test_memory_lifetime.cc  11
- src/vt/worker/worker_group.h  1
- src/vt/parameterization/parameterization.h  2
- src/vt/trace/trace.h  2
- src/vt/runtime/runtime.cc  2
- src/vt/messaging/active.cc  2
- src/vt/pool/pool.cc  2
- src/vt/sequence/sequencer.impl.h  3
- src/vt/rdma/rdma.h  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5390178&bid=17670967)

**Comment by lifflander** [comment]
@PhilMiller @nlslatt @pnstickne  Please review!

**Comment by lifflander** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- tests/unit/memory/test_memory_lifetime.cc  11
- src/vt/trace/trace.h  2
- src/vt/pool/pool.cc  2
- src/vt/worker/worker_group_omp.h  1
- src/vt/sequence/sequencer.impl.h  3
- src/vt/rdma/rdma.h  6
- src/vt/worker/worker_group.h  1
- src/vt/messaging/active.h  4
- src/vt/parameterization/parameterization.h  2
- src/vt/messaging/active.cc  2
- src/vt/runtime/runtime.cc  2
- tests/unit/memory/test_memory_active.cc  2
- src/vt/trace/trace.cc  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/lifflander/vt/pullRequest?prid=5390178&bid=17670967)

File: src/vt/runtime/component/component.h [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+   * \brief Empty default overridden progress method
+   *
+   * \return no units processed
+   */
+  virtual int progress() override { return 0; }
+
+  /**
+   * \brief Empty default diagnostic dump state
+   */
+  void dumpState() override { }
```
Comment (pnstickne) [inline src/vt/runtime/component/component.h:150]: `virtual`?

File: src/vt/runtime/component/component_pack.impl.h [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+namespace vt { namespace runtime { namespace component {
+
+template <typename T, typename... Deps, typename... Cons>
+registry::AutoHandlerType ComponentPack::registerComponent(
+  T** ref, typename BaseComponent::DepsPack<Deps...>, Cons&&... cons
+) {
+  ComponentRegistry::dependsOn<T, Deps...>();
+  auto idx = registry::makeIdx<T>();
+  registered_components_.push_back(idx);
+  construct_components_.emplace(
```
Comment (pnstickne) [inline src/vt/runtime/component/component_pack.impl.h:59]: Would separate out [nested] creation and adding to components.

File: src/vt/runtime/runtime_component_fwd.h [hunk 3]
```diff
@@ -97,6 +97,9 @@ struct PipeManager;
 namespace objgroup {
 struct ObjGroupManager;
 }
+namespace util { namespace memory {
```
Comment (pnstickne) [inline src/vt/runtime/runtime_component_fwd.h:100]: New line for consistency.

File: src/vt/trace/trace.cc [hunk 4]
```diff
@@ -172,52 +171,48 @@ bool Trace::inIdleEvent() const {
...
-  std::string const& in_prog_name, std::string const& in_trace_name,
-  std::string const& in_dir_name
-) {
-  auto const node = theContext()->getNode();
-
-  prog_name_ = in_prog_name;
-  trace_name_ = in_trace_name;
-  start_time_ = getCurrentTime();
+void Trace::setupNames(std::string const& in_prog_name) {
+  if (ArgType::vt_trace) {
```
Comment (pnstickne) [inline src/vt/trace/trace.cc:175]: Might be cleaner to if-not-return.

File: src/vt/objgroup/manager.h [hunk 5]
```diff
@@ -81,6 +82,13 @@ struct ObjGroupManager {
 
   ObjGroupManager() = default;
 
+  /**
```
Comment (pnstickne) [inline src/vt/objgroup/manager.h:85]: I'd be fine with removing these comments, as they are not applied consistently.
Then unified comments can be added to the base type(s).

File: src/vt/parameterization/parameterization.h [hunk 6]
```diff
@@ -116,7 +117,14 @@ static void dataMessageHandler(DataMsg<Tuple>* msg) {
 #endif
 }
 
-struct Param {
+struct Param : runtime::component::Component<Param> {
+
+  /**
```
Comment (pnstickne) [inline src/vt/parameterization/parameterization.h:122]: Ditto.

File: src/vt/runtime/component/component.h [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+  static std::unique_ptr<T> staticInit(Args&&... args, hasNoCons<U>* = nullptr) {
+    return std::make_unique<T>(std::forward<Args>(args)...);
+  }
+
+  /**
+   * \brief Whether the component requires the scheduler to poll
+   *
+   * \return default component is not pollable
+   */
+  bool pollable() override {
```
Comment (pnstickne) [inline src/vt/runtime/component/component.h:121]: Add unified comments to base/virtual and allow inherit?

File: src/vt/runtime/component/component_pack.cc [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+*/
+
+#if !defined INCLUDED_VT_RUNTIME_COMPONENT_COMPONENT_PACK_CC
+#define INCLUDED_VT_RUNTIME_COMPONENT_COMPONENT_PACK_CC
+
+#include "vt/runtime/component/component_pack.h"
+
+namespace vt { namespace runtime { namespace component {
+
+void ComponentPack::construct() {
```
Comment (pnstickne) [inline src/vt/runtime/component/component_pack.cc:52]: While neat, might be overkill as there is a fixed set of VT deps, and VT is itself not a dependency management library to clients.

File: src/vt/runtime/component/component.h (hunk 1 above)
Comment (lifflander) [reply src/vt/runtime/component/component.h:150]: Fixed

File: src/vt/trace/trace.cc (hunk 4 above)
Comment (lifflander) [reply src/vt/trace/trace.cc:175]: Done

File: src/vt/objgroup/manager.h (hunk 5 above)
Comment (lifflander) [reply src/vt/objgroup/manager.h:85]: Fixed

File: src/vt/runtime/component/component_pack.cc (hunk 8 above)
Comment (lifflander) [reply src/vt/runtime/component/component_pack.cc:52]: I think as we add more components to VT it might be useful. Also, there was a lot of tricky code before to get the startup order correct. This solves it in a uniform way

**Comment by lifflander** [comment]
@PhilMiller Please review!

**Comment by PhilMiller** [comment]
In progress

File: src/vt/runtime/component/progressable.h [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+#if !defined INCLUDED_VT_RUNTIME_COMPONENT_PROGRESSABLE_H
+#define INCLUDED_VT_RUNTIME_COMPONENT_PROGRESSABLE_H
+
+#include "vt/config.h"
+
+namespace vt { namespace runtime { namespace component {
+
+struct Progressable {
+  virtual int progress() = 0;
```
Comment (PhilMiller) [inline src/vt/runtime/component/progressable.h:57]: This needs a comment to document the meaning of the return value.

File: src/vt/runtime/runtime.cc [hunk 10]
```diff
@@ -1190,62 +1189,221 @@ void Runtime::finalizeContext() {
...
+  p.add<param::Param>();
+  p.add<seq::Sequencer>();
+  p.add<seq::SequencerVirtual>();
+  p.add<location::LocationManager>();
+  p.add<vrt::VirtualContextManager>();
+  p.add<vrt::collection::CollectionManager>();
+  p.add<rdma::Manager>();
+  p.add<registry::Registry>();
+  p.add<event::AsyncEvent>();
+  p.add<pool::Pool>();
```
Comment (PhilMiller) [inline src/vt/runtime/runtime.cc]: Why are there separate `registerComponent` and `add` calls? These seem redundant.

Also, there's at least one inconsistency between `rdma::Manager` and `rdma::RDMAManager`

File: src/vt/runtime/runtime.cc [hunk 11]
```diff
@@ -1012,9 +1012,7 @@ bool Runtime::finalize(bool const force_now) {
     fflush(stdout);
     fflush(stderr);
     sync();
-    finalizeComponents();
-    finalizeOptionalComponents();
-    finalizeTrace();
+    p_ = nullptr;
```
Comment (PhilMiller) [inline src/vt/runtime/runtime.cc:1020]: Stylistically, I think it's slightly preferable to say `p_.reset(nullptr)`, but that's a nitpick

File: src/vt/runtime/runtime.h [hunk 12]
```diff
@@ -180,7 +180,7 @@ struct Runtime {
   bool runtime_active_ = false;
   bool is_interop_ = false;
   WorkerCountType num_workers_ = no_workers;
-  MPI_Comm* communicator_ = nullptr;
```
Comment (PhilMiller) [inline src/vt/runtime/runtime.h:187]: I don't think there's actually a need to handle an `MPI_Comm` as a pointer - they're typically already handles to an MPI-internal data structure, and so it makes sense to pass them around by value.

If you were doing some RAII thing of duplicating on construction, and freeing on destruction, this would be a bit more sensible. I don't think there's a need for that in this circumstance.

Comment (PhilMiller) [reply src/vt/runtime/runtime.h:187]: There's a sentinel value `MPI_COMM_NULL` defined in the standard.

File: src/vt/runtime/runtime.h [hunk 13]
```diff
@@ -81,6 +81,7 @@ struct Runtime {
 
   virtual ~Runtime();
 
+  bool live() const { return p_ and p_->live(); }
```
Comment (PhilMiller) [inline src/vt/runtime/runtime.h:84]: This should be named `isLive()`

**Comment by lifflander** [comment]
> It would be wise, I think, for the component interface to have a method `std::string getComponentName()`, so that debug prints during startup and shutdown can use that automatically, and less needs to be baked into each individually

I don't think I understand. How do we automatically get the name without having an overload per component? typeid?

File: src/vt/runtime/component/progressable.h (hunk 9 above)
Comment (lifflander) [reply src/vt/runtime/component/progressable.h:57]: Added

File: src/vt/runtime/runtime.cc (hunk 10 above)
Comment (lifflander) [reply src/vt/runtime/runtime.cc]: `registerComponent` does static registration with dependencies, but does not add it to the construct set. When a component is added, it will be constructed along with all its dependencies. 

I realize that the distinction isn't needed right now, but thought it might be useful in the future. Say you want a customized runtime with just collection functionality, you could just add collection and get that plus everything it needs to run properly.

Regarding `rdma::Manager` vs `rdma::RDMAManager` that's an unfortunate results of having two very different rdma components in the system right now. Eventually, that will get resolved.

File: src/vt/runtime/runtime.cc (hunk 11 above)
Comment (lifflander) [reply src/vt/runtime/runtime.cc:1020]: Done

File: src/vt/runtime/runtime.h (hunk 13 above)
Comment (lifflander) [reply src/vt/runtime/runtime.h:84]: Done

File: src/vt/runtime/runtime.h (hunk 12 above)
Comment (lifflander) [reply src/vt/runtime/runtime.h:187]: I can get rid of the unique_ptr there. The reason it was done is because it gets potentially modified by the context, but I can just pass the memory address.

Comment (lifflander) [reply src/vt/runtime/runtime.h:187]: Addressed

File: src/vt/messaging/active.cc [hunk 14]
```diff
@@ -905,7 +905,7 @@ bool ActiveMessenger::testPendingDataMsgAsyncRecv() {
   );
 }
 
-bool ActiveMessenger::progress() {
+int ActiveMessenger::progress() {
```
Comment (PhilMiller) [inline src/vt/messaging/active.cc:908]: This is probably OK, but it isn't strictly satisfying the contract of a "count" of units of progress made.

File: src/vt/worker/worker_group_omp.cc [hunk 15]
```diff
@@ -81,8 +81,9 @@ bool WorkerGroupOMP::commScheduler() {
   return WorkerGroupComm::schedulerComm(finished_fn_);
 }
 
-void WorkerGroupOMP::progress() {
+int WorkerGroupOMP::progress() {
   WorkerGroupCounter::progress();
+  return 0;
```
Comment (PhilMiller) [inline src/vt/worker/worker_group_omp.cc:86]: Is it safe for a component to uniformly return 'no progress'?

File: src/vt/vrt/collection/manager.h [hunk 16]
```diff
@@ -90,7 +91,9 @@
 
 namespace vt { namespace vrt { namespace collection {
 
-struct CollectionManager {
+struct CollectionManager
+  : runtime::component::PollableComponent<CollectionManager>
```
Comment (PhilMiller) [inline src/vt/vrt/collection/manager.h:95]: Should this be a non-pollable `Component`, since `progress` always returns 0?

File: src/vt/trace/trace.cc [hunk 17]
```diff
@@ -172,17 +171,15 @@ bool Trace::inIdleEvent() const {
...
+  }
 
-  prog_name_ = in_prog_name;
-  trace_name_ = in_trace_name;
-  start_time_ = getCurrentTime();
+  auto const node = theContext()->getNode();
 
-  std::string dir_name = (in_dir_name.empty()) ? prog_name_ + "_trace" : in_dir_name;
+  trace_name_ = prog_name_ + "." + std::to_string(node) + ".log.gz";
+  auto dir_name = prog_name_ + "_trace";
```
Comment (PhilMiller) [inline src/vt/trace/trace.cc:182]: I'm not sure why this change of logic is coming with this PR, but it doesn't seem objectionable as such.

File: src/vt/sequence/sequencer_fwd.h [hunk 18]
```diff
@@ -50,15 +50,13 @@
 
 namespace vt { namespace seq {
 
-template <typename SeqTag, template <typename> class SeqTrigger>
-struct TaggedSequencer;
+struct Sequencer;
 
-using Sequencer = TaggedSequencer<SeqType, SeqMigratableTriggerType>;
+//using Sequencer = TaggedSequencer<SeqType, SeqMigratableTriggerType>;
```
Comment (PhilMiller) [inline src/vt/sequence/sequencer_fwd.h:55]: Why not just delete these, rather than comment them out?

**Comment by PhilMiller** [comment]
> > It would be wise, I think, for the component interface to have a method `std::string getComponentName()`, so that debug prints during startup and shutdown can use that automatically, and less needs to be baked into each individually
> 
> I don't think I understand. How do we automatically get the name without having an overload per component? typeid?

Never mind, I see that a later commit actually added a `name` method. That's all I was asking for.

My original comment was made as I was reviewing commit-by-commit, since they seemed to form a decent progression.

**Comment by PhilMiller** [comment]
Assuming none of my last few comments indicate a real issue to you, this has my +1

File: src/vt/trace/trace.cc (hunk 17 above)
Comment (lifflander) [reply src/vt/trace/trace.cc:182]: It used to be in the runtime and I had to move it over to Trace where it should be now that Runtime avoids component-specific logic.

File: src/vt/messaging/active.cc (hunk 14 above)
Comment (lifflander) [reply src/vt/messaging/active.cc:908]: Yes, right now the return value is not used. Eventually it might be, when that happens we'll make all these return the correct values.

File: src/vt/worker/worker_group_omp.cc (hunk 15 above)
Comment (lifflander) [reply src/vt/worker/worker_group_omp.cc:86]: Yes, it is. Eventually we can make these return correctly if needed.

File: src/vt/vrt/collection/manager.h (hunk 16 above)
Comment (lifflander) [reply src/vt/vrt/collection/manager.h:95]: Done, fixed

File: src/vt/sequence/sequencer_fwd.h (hunk 18 above)
Comment (lifflander) [reply src/vt/sequence/sequencer_fwd.h:55]: My mistake
